### PR TITLE
MOO-380: dhis2 integration

### PR DESCRIPTION
**Pull request by kaweesi**

### CODE REVIEW

**Comment by tmvumbi2** [comment]
Nice work @kaweesi, below is a feedback from functional testing:
- the property file `dhis-mappinngs.properties` and the global variable `msfcore.dhis.ncdProgramUid` did not have the correct values after installing msfcore module with dhis2 integration code (needed to update them manually)
-  when testing patient enrollment, the attribute 'nationality' is causing the sync to break (error: `Value 'brazil' is not a valid option for attribute Xvzc9e0JJmp and option set qr9jBtm9uvm`): this is likely due to the fact not all countries have corresponding option sets in DHIS2 (see an example of a breaking request here: http://146.185.183.146:9000/#!/transactions/5c34ba5b30d100001195d634)
- when testing patient enrollment, the following fields listed in `dhis-mapping.properties` were not mapped: sex, education, employment status, patient type, refugee status, type of malnutrition, cause of referral, initials of case manager (the last 5 fields seems not to be part of msfcore patient registration; if confirmed, should we also remove them from the mapping file?). Note: I tested by creating a new patient with all the fields populated (see the transaction here:  http://146.185.183.146:9000/#!/transactions/5c35da7c30d100001195f966)
- when testing patient enrollment, a patient record has been created 3 times in DHIS2 during sync (search for 'Glody' in DHIS2)
- sync of baseline: I created a baseline form with all the fields from all the sections populated, the payload sent to OpenHIM contained the 2 following data elements: tQbAznqZnCr (value: '2019-01-24': this corresponds to the request appointment date) and Yt2KI46b58O (value: 'Comments': this corresponds to the lifestyle comment). However, the previous IDs are associated to different fields in dataElements.json: `tQbAznqZnCr` = `Social Work – Date`  and `Yt2KI46b58O` = `Surgery – EXT- Comments`. In DHIS2 (see http://146.185.183.146/dhis-web-tracker-capture/index.html#/dashboard?tei=w6pz4CNvBmL&program=l3r2cXKdmuv&ou=izBEzgovieF&fromAudit), the above values do not appear under the baseline section. See the message sent to DHIS here: http://146.185.183.146:9000/#!/transactions/5c35bc0a30d100001195f52e
- sync of follow up: I created a follow up form with all the fields from all the sections populated, the payload sent to OpenHIM contained one data element: tQbAznqZnCr (value: '2019-01-26': this corresponds to the request appointment date). However, the id tQbAznqZnCr is associated to a different field in dataElements.json: `tQbAznqZnCr` = `Social Work – Date`. In DHIS2 (see: http://146.185.183.146/dhis-web-tracker-capture/index.html#/dashboard?tei=w6pz4CNvBmL&program=l3r2cXKdmuv&ou=izBEzgovieF&fromAudit), the value associated to tQbAznqZnCr does not appear under the follow up section. See the message sent to DHIS here: http://146.185.183.146:9000/#!/transactions/5c35c07030d100001195f5e9

File: api/src/main/resources/sync2.json [hunk 1]
```diff
@@ -123,7 +123,13 @@
         "category" : "audit_message",
         "openMrsClass": "org.openmrs.module.sync2.api.model.audit.AuditMessage",
         "enabled" : false
-      }
+      },
+      {
+          "classTitle" : "DHISReporting",
+          "category" : "dhis_eporting",
```
Comment (tmvumbi2) [inline api/src/main/resources/sync2.json]: @kaweesi small typos: `dhis_reporting` (missing 'r')

File: api/src/main/java/org/openmrs/module/msfcore/api/impl/DHISServiceImpl.java [hunk 2]
```diff
@@ -302,22 +330,88 @@ private void addTrackInstanceAttributes(Patient patient, Properties dhisMappings
...
+                    }
+                }
+            }
+            data.getDataElements().putAll(dataElements);
+        }
+
         parameters.setEventDate(today);
     }
 
+    private String getDHISValueFromObs(Obs obs, SimpleJSON optionSets) {
```
Comment (tmvumbi2) [inline api/src/main/java/org/openmrs/module/msfcore/api/impl/DHISServiceImpl.java]: @kaweesi optionSets is not used in this method

**Comment by tmvumbi2** [comment]
@kaweesi 

DHISReporting and Liquibase.xml:
- about the table's name: this table contains a list of patients, and their DHIS sync status. I would suggest to rename this table to something like: PatientDHISSyncStatus.
- I would suggest to replace the fields stage_changed and stage_edited with a single field "syncStatus" which can have the following values: NO_SYNC, ENROLLMENT and STAGE_CHANGE. This change will simplify the table and the implementation of methods such as triggerOnDHISReportingStageChanged and triggerOffDHISReporting. This field can be an enum (i.e. DHISSyncStatus with values: NO_SYNC, ENROLLMENT, STAGE_CHANGE).
- I'd suggest to remove the method getDHISRequest() and the enum DHISRequest: the getter for the field syncStatus can do the job here
- I'd suggest to rename pending() to isPending()
- I'd suggest to rename trackedEntityInstance to trackedEntityInstanceId
- I'd suggest to rename getPatient() to getPatientUuid()
- removing the foreign key for the patient uuid makes it possible to refer to a non-existing patient (which might create problems during sync). 
- Comment about the usage of the patient uuid: there is a performance side effect created with using the uuid: text columns are not effecient to be used as foreign keys. I'm not sure about a way around this for now.

DHISService: I'd suggest the following method renaming:
- postPendingTrackerInstancesThroughOpenHIM: syncPatientsToDHIS()
- postTrackerInstanceThroughOpenHimForAPatient: syncSinglePatientToDHIS()
- triggerOnDHISReportingStageChanged: flagPatientForDHISSync()
- triggerOffDHISReporting: unflagPatientForDHISSync()
- getPendingDHISReportings: getPatientsToSyncToDHIS()
- setPendingDHISReportings: setPatientDHISSyncStatus()

DHISServiceImpl:
- I'd suggest to extract all the logic to build a TrackerInstance into an helper class (i.e. OpenMRSToDHIS), this will simplify DHISServiceImpl
- postPendingTrackerInstancesThroughOpenHIM: I'd suggest to extract the condition on line 494 (checking if the current location is a project) into a private method
- getDHISEndPointFile(): this method creates a folder ([APP_DIR]/msfcore) if it doesn't exist, before returning a File's instance. I'd suggest to rename the method to createDHISEndPointFile(...). And also rename msfcoreStorage to msfcoreFolder
- I would sugest to replace the generic SimpleJSON by a custom data structure (based on a simple POJO for example); the reason behind this is that SimpleJSON isn't simple to navigate through (see for ex the implementation of OpenMRSToDHIS.getMatchedDhisFieldByFieldAndValue(...))
- installDHIS2Metadata and updateDHIS2Metadata look similar, is there a way to have a single method that does both the first install and update?
- dhisNcdTransitioning: this method intent isn't clear from the name (would suggest to use an active verb on the name that summarizes what the method is doing). I'd also suggest not to change the value of a parameter (i.e. like done for trackerInstance; see https://softwareengineering.stackexchange.com/questions/245767/is-modifying-an-incoming-parameter-an-antipattern). Fixing this would require to rethink a little the class design to avoid this pattern.
- setProgramStageAndEncounters: the method acts differently when enterTypeUuid parameter is provided. I'd suggest to create a separate method to handle the case of request investigation; and since the encounterTypeUuid is known, it can be hard coded into the method (i.e. not passed as a parameter).
- setProgramStageAndEncounters:  the implementation of this method is opaque and fairly hard to read, I'd suggest to break it down into smaller steps with clear variable names, to make the code more readable

MSFCoreConfig.java
- WORKFLOW_STATE_UUID_INVESTIGATION_RESULTS & GP_DHIS_NCD_PROGRAMSTAGE_UID_INVESTIGATION_RESULTS: from the code, it looks like a stage change is triggered when a request investigation is captured. If this is the case, I'd suggest to re-name the 'INVESTIGATION_RESULTS' part into 'REQUEST_INVESTIGATION' (the current names suggest that the stage change is triggered when the investigation result is recorded). See also INVESTIGATION_RESULTS in ProgramStages.java and concept_name concept_name_id="500009" name="Investigation results" in MSFCustomConcepts.xml

ConfigurationPageController.java:
- needs tests

OpenMRSToDHIS:
- I'd suggest to rename the class to OpenMRSToDHISMapper
- getCodeFromConceptReferenceTerm: please add a FIXME (or TODO?) to signal that this method needs to be implemented

Other questions:
- what should happen when a patient is recorded with a baseline form containing a request investigation? Both the baseline and investigation stage should be created in DHIS?
- PatientTrackableAttributes: Why are mother's and father's names removed from this enum?
- TrackedEntityAttributes.json: why is this file needed? When sending data to DHIS, the attribute ids are used  and they are mapped in dhis-mapping.properties

**Comment by kaweesi** [comment]
Thanks for the suggestions @tmvumbi2, i have quickly perused through your feedback above

- About the removal of DHISRequest, we send that as part of the payload from OpenMRS, in your suggestion, DHISSyncStatus is probably usable.
- patient_uuid is no-longer a foreign key anymore i think.
- TrackedEntityAttributes.json is used when enrolling, it contains patient entity attributes. 
- about most other refactoring, i think we can consider the amount of work and urgency or most importantly runtime improvement when we resume working on MSF.

Most of the work in this PR has been evolving since the requirements weren't clear from when we started off and have been progressing on how to implement the requirements which brought us to where we are, i believe we can improve a lot of this with more requirements knowledge.

**Comment by kaweesi** [comment]
https://jembiprojects.jira.com/browse/MOO-903 resolves a number of those reviews @tmvumbi2

**Pull request merged**